src/server/pps/server/pipeline_state_driver.go

```diff
@@ -0,0 +1,272 @@
+	if err := setPipelineState(ctx, sd.db, sd.pipelines, specCommit, state, reason); err != nil {
+		// don't bother failing if we can't set the state
+		return stepError{
+			error: errors.Wrapf(err, "could not set pipeline state to %v"+
+				"(you may need to restart pachd to un-stick the pipeline)", state),
+			retry: true,
+		}
+	}
+	return nil
+}
+
+func (sd *stateDriver) TransitionState(ctx context.Context, specCommit *pfs.Commit, from []pps.PipelineState, to pps.PipelineState, reason string) (retErr error) {
+	span, ctx := tracing.AddSpanToAnyExisting(ctx,
+		"/pps.Master/TransitionPipelineState", "pipeline", specCommit.Branch.Repo.Name,
+		"from-state", from, "to-state", to)
+	defer func() {
+		tracing.TagAnySpan(span, "err", retErr)
+		tracing.FinishAnySpan(span)
+	}()
+	return ppsutil.SetPipelineState(ctx, sd.db, sd.pipelines,
+		specCommit, from, to, reason)
+}
+
+func (sd *stateDriver) Watch(ctx context.Context) (<-chan *watch.Event, func(), error) {
+	pipelineWatcher, err := sd.pipelines.ReadOnly(ctx).Watch()
+	if err != nil {
+		return nil, nil, errors.EnsureStack(err)
+	}
+	return pipelineWatcher.Watch(), pipelineWatcher.Close, nil
+}
+
+func (sd *stateDriver) ListPipelineInfo(ctx context.Context, f func(*pps.PipelineInfo) error) error {
+	return ppsutil.ListPipelineInfo(ctx, sd.pipelines, nil, 0, f)
+}
+
+func (sd *stateDriver) GetPipelineInfo(ctx context.Context, name string, version int) (*pps.PipelineInfo, error) {
+	var pipelineInfo pps.PipelineInfo
+	if err := sd.pipelines.ReadOnly(ctx).GetUniqueByIndex(
+		ppsdb.PipelinesVersionIndex,
+		ppsdb.VersionKey(name, uint64(version)),
+		&pipelineInfo); err != nil {
+		return nil, errors.Wrapf(err, "couldn't retrieve pipeline information")
+	}
+	return &pipelineInfo, nil
+}
+
+func (sd *stateDriver) tryLoadLatestPipelineInfo(ctx context.Context, pipeline string) (*pps.PipelineInfo, error) {
+	pi := &pps.PipelineInfo{}
+	errCnt := 0
+	err := backoff.RetryNotify(func() error {
+		return sd.loadLatestPipelineInfo(ctx, pipeline, pi)
+	}, backoff.NewExponentialBackOff(), func(err error, d time.Duration) error {
+		errCnt++
+		// Don't put the pipeline in a failing state if we're in the middle
+		// of activating auth, retry in a bit
+		if (auth.IsErrNotAuthorized(err) || auth.IsErrNotSignedIn(err)) && errCnt <= maxErrCount {
+			log.Warnf("PPS master: could not retrieve pipelineInfo for pipeline %q: %v; retrying in %v",
+				pipeline, err, d)
+			return nil
+		}
+		return stepError{
+			error: errors.Wrapf(err, "could not load pipelineInfo for pipeline %q", pipeline),
+			retry: false,
+		}
+	})
+	return pi, err
+}
+
+func (sd *stateDriver) loadLatestPipelineInfo(ctx context.Context, pipeline string, message *pps.PipelineInfo) error {
+	specCommit, err := ppsutil.FindPipelineSpecCommit(ctx, sd.pfsApi, *sd.txEnv, pipeline)
+	if err != nil {
+		return errors.Wrapf(err, "could not find spec commit for pipeline %q", pipeline)
+	}
+	// TODO: which context should be passed here? master's ctx?
```

Comment:
what situations are you thinking of where the request/pipeline controller context wouldn't be appropriate? We're already a few function calls deep at this point, I guess

Reply:
I actually think this TODO is outdated. The function should be passed the PC's ctx